fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -85,122 +42,629 @@ def dloss(input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
+        return self
+
+    def get_module_rref(self) -> rpc.RRef:
+        return self.module_rref
+
+
+class PipelineModulesGraph(nn.Module):
+    """A collection of remote modules (of type PipelineModule) with connections showing how inputs
+    to the model or outputs of individual modules are use as inputs of subsequent modules.
+    The graph has a number of helper functions that add new modules to the graph and define inputs
+    to these module.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.modules_list: List = []
+        # self.inputs specifies inputs to each module in modules_list. Each input to each module
+        # is represented by a tuple (i, j). If i>=0, then the input is the j'th output of the i'th
+        # module. If i<0, the input is the j'th input to the model.
+        self.inputs: List[Optional[List[Tuple[int, int]]]] = []
+
+    def _find_or_add(self, module: PipelineModule) -> int:
+        try:
+            return self.modules_list.index(module)
+        except ValueError:
+            self.inputs.append(None)
+            self.modules_list.append(module)
+            return len(self.modules_list) - 1
+
+    def add_sequence(self, modules: List[PipelineModule], first_input: Optional[PipelineModule] = None) -> None:
+        """Adds a list of modules to the graph, to be run sequentially.
+        The connection between these modules is as follows: the first output of each of these modules
+        (except the last one) is used as the first input of its next module in this sequence.
+        The user may also specify the input to the first module in this sequence with argument 'first_input'.
+        In this case the module 'first_input' must have been added to the graph previously.
+        """
+        old_modules_len = len(self.modules_list)
+        new_modules_len = len(modules)
+        self.modules_list.extend(modules)
+        # update inputs array
+        self.inputs.append([(self.modules_list.index(first_input), 0)] if first_input is not None else None)
+        for i in range(old_modules_len + 1, old_modules_len + new_modules_len):
+            self.inputs.append([(i - 1, 0)])
+
+    def feed_model_input(self, module: PipelineModule, ind: int = 0) -> None:
+        """Declares the input to a module as the input to the model. In case the model has multiple
+        inputs, the argument 'ind' indicates the index of the model input that is fed to the module.
+        """
+        self.inputs[self._find_or_add(module)] = [(-1, ind)]
+
+    def add_multi_input_layer(self, module: PipelineModule, inputs: List[PipelineModule]) -> None:
+        """Adds a module with multiple inputs to the graph. The modules that provide inpurs to this module
+        must have been added previously to the graph and are listed with argument inputs.
+        """
+        self.inputs[self._find_or_add(module)] = [(self.modules_list.index(m), 0) for m in inputs]
```

Comment:
why is it always the 0th output?

Reply:
The API for now does not support other outputs. In next iterations we can extend the signature to accept something like this:
```py 
inputs: List[Union[PipelineModule, Tuple[PipelineModule, int]]]
```
to enable user to set output index if needed.